x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtUtil.java

```diff
@@ -0,0 +1,312 @@
+    ) throws SettingsException {
+        switch (clientAuthenticationType) {
+            case JwtRealmSettings.CLIENT_AUTHENTICATION_TYPE_SHARED_SECRET:
+                // If type is "SharedSecret", the shared secret value must be set
+                if (Strings.hasText(clientAuthenticationSharedSecret) == false) {
+                    throw new SettingsException(
+                        "Missing setting for ["
+                            + clientAuthenticationSharedSecretConfigKey
+                            + "]. It is required when setting ["
+                            + clientAuthenticationTypeConfigKey
+                            + "] is ["
+                            + JwtRealmSettings.CLIENT_AUTHENTICATION_TYPE_SHARED_SECRET
+                            + "]"
+                    );
+                }
+                break;
+            case JwtRealmSettings.CLIENT_AUTHENTICATION_TYPE_NONE:
+            default:
+                // If type is "None", the shared secret value must not be set
+                if (Strings.hasText(clientAuthenticationSharedSecret)) {
+                    throw new SettingsException(
+                        "Setting ["
+                            + clientAuthenticationSharedSecretConfigKey
+                            + "] is not supported, because setting ["
+                            + clientAuthenticationTypeConfigKey
+                            + "] is ["
+                            + JwtRealmSettings.CLIENT_AUTHENTICATION_TYPE_NONE
+                            + "]"
+                    );
+                }
+                LOGGER.warn(
+                    "Setting ["
+                        + clientAuthenticationSharedSecretConfigKey
+                        + "] value ["
+                        + JwtRealmSettings.CLIENT_AUTHENTICATION_TYPE_NONE
+                        + "] may not be secure. Unauthorized clients may be able to submit JWTs from the same issuer."
+                );
+                break;
+        }
+    }
+
+    public static void validateClientAuthentication(final String type, final SecureString expectedSecret, final SecureString actualSecret)
+        throws Exception {
+        switch (type) {
+            case JwtRealmSettings.CLIENT_AUTHENTICATION_TYPE_SHARED_SECRET:
+                if (Strings.hasText(actualSecret) == false) {
+                    throw new Exception("Rejected client authentication for type [" + type + "] due to no secret.");
+                } else if (expectedSecret.equals(actualSecret) == false) {
+                    throw new Exception("Rejected client authentication for type [" + type + "] due to secret mismatch.");
+                }
+                break;
+            case JwtRealmSettings.CLIENT_AUTHENTICATION_TYPE_NONE:
+            default:
+                if (Strings.hasText(actualSecret)) {
+                    throw new Exception("Rejected client authentication for type [" + type + "] due to present secret.");
```

Comment:
If you look at the setting as “what we’re validating” it’s 

None => We validate nothing no matter what you send
SharedSecret => We validate the shared secret if we find the header , otherwise we fail
PKI => We validate client cert chain if we get one, otherwise we fail 

inthibk this makes sense